### ISSUE TITLE
Could not download OLCI from https://coda.eumetsat.int

### ISSUE THREAD

**Issue by paalge** [sentinelsat/sentinelsat#311]
Hi

I'm new to using sentinelsat.
I'm currently running the master branch (v0.13-37-g533e243) of sentinelsat (due to this issue #294)  and trying to download OLCI data from eumetsat.
The command is this:

`sentinelsat --url https://coda.eumetsat.int  -u uname -p 'password' -g some.json --sentinel 3 -s 20190515 -e 20190518 --instrument OLCI  -d`

The output I get is this:

Found 25 products
Will download 25 products using 2 workers
```
Traceback (most recent call last):
  File "/home/pal/.local/bin/sentinelsat", line 11, in <module>
    load_entry_point('sentinelsat==0.13', 'console_scripts', 'sentinelsat')()
  File "/home/user/.local/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/user/.local/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/user/.local/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/user/.local/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/user/.local/lib/python3.6/site-packages/sentinelsat/scripts/cli.py", line 168, in cli
    product_infos, triggered, failed_downloads = api.download_all(products, path)
  File "/home/user/.local/lib/python3.6/site-packages/sentinelsat/sentinel.py", line 690, in download_all
    raise next(iter(x.exception() for x in dl_tasks if x.exception()))
  File "/usr/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/user/.local/lib/python3.6/site-packages/sentinelsat/sentinel.py", line 761, in _download_online_retry
    if self.is_online(product_info['id']):
  File "/home/user/.local/lib/python3.6/site-packages/sentinelsat/sentinel.py", line 457, in is_online
    raise SentinelAPIError('Could not verify whether product {} is online'.format(id), r)
sentinelsat.sentinel.SentinelAPIError: HTTP status 404 Not Found: Could not verify whether product 4c6186dd-f89e-4164-86bb-b597a478821a is online
```

The json file is a 5 point polygon.

Is this a bug, or a misstake on my side?

**Comment by j08lue** [comment]
> Is this a bug, or a misstake on my side?

No, probably just another hub / case that behaves differently from SciHub regarding the Long-Term Archive functionality. 

As a fast fix (before we get to fixing this properly) you can replace this line with `return True`

https://github.com/sentinelsat/sentinelsat/blob/80d01bdf28f5330e5515a6d77f4159dd0194198d/sentinelsat/sentinel.py#L457

**Comment by paalge** [comment]
Replacing the line fixed the issue. 
Thank you

**Comment by j08lue** [comment]
For our own records: This is not fixed yet.

**Comment by gbaier** [comment]
@paalge could you try wether you can access the **online** property value directly using https://coda.eumetsat.int.

Something like this should return true if the product is online: https://coda.eumetsat.int/dhus/odata/v1/Products('98ca202b-2155-4181-be88-4358b2cbaaa0')/Online/$value

See also https://scihub.copernicus.eu/twiki/do/view/SciHubUserGuide/LongTermArchive#Management_of_offline_products_v concerning what kind of behaviour sentinelsat expects from a hub.

**Comment by paalge** [comment]
I get not available
https://coda.eumetsat.int/dhus/odata/v1/Products('07d54fc9-66d3-4f85-a06d-5b69ee7a7ccf')/Online/$value
though I'm not sure that is what you were looking for

**Comment by gbaier** [comment]
Sorry for the long silence.

Since the issues seems to be that some hubs apparently do not fully implement the LTA API, how about changing the internals from `is_online` to account for that.

If `is_online` just called `get_product_odata` and evaluate the **Online** flag this should solve the issue. My only gripe would be that a lot of information, i.e. the whole odata, would be regularly requested, which seems to be a little bit wasteful.

Anyways, corrections or opinions are welcome ;)

**Comment by j08lue** [comment]
> My only gripe would be that a lot of information, i.e. the whole odata, would be regularly requested, which seems to be a little bit wasteful.

We could first try to use the LTA API and use odata as a fallback option.

**Comment by Fernerkundung** [comment]
I sent a mail to the Copernicus support asking about how this should be handled on the user side.

In theory these should be handled identically since all the hubs use the same software. It looks like these LTAs are implemented differently depending on the hub. 

I have yet to receive an answer from the support.

**Comment by mstrahl** [comment]
Just to consider that the issue stems mostly from what DhuS version is running on each hub. So maybe the new code with LTA support could be run for newer DhuS versions and older versions fall back to the old code that worked well for them.
Cheers
Mikko

**Comment by Fernerkundung** [comment]
Any suggestions on how we can find out the DHuS version or the LTA policy programmatically?

https://scihub.copernicus.eu/dhus/odata/v1/$metadata does not seem to be very helpful in this regard.

**Comment by mstrahl** [comment]
Hi

Maybe not from the odata interface, but the DHuS website has somewhere the
version info.

Cheers
Mikko

su 26. tammik. 2020 klo 17.48 Kersten <notifications@github.com> kirjoitti:

**Comment by mstrahl** [comment]
https://finhub.nsdc.fmi.fi/#/home and https://colhub.met.no/#/home have on their front page without having to login an info button and a related <div id="datahubversion">
Text on finhub is:
 Finnish Data Hub System 0.13.4-21-1 developed by a Serco and GAEL Systems consortium under a contract with the European Space Agency - Funded by the EU and ESA
and colhub:
Data Hub System 0.13.4-22 developed by a Serco and GAEL Systems consortium under a contract with the European Space Agency - Funded by the EU and ESA

so I guess it is fairly certain to read the string Data Hub System and related version number from there.

Cheers
Mikko

**Comment by gbaier** [comment]
I did some digging how to get the version info, but I'm also stuck.

The text in the info string, except the version info, can be accessed via https://scihub.copernicus.eu/appconfig.json.

dhus info can be queried using URLs like https://scihub.copernicus.eu/dhus/odata/v1/Users. Accessing https://scihub.copernicus.eu/dhus/odata/v1/$metadata returns and XML with what exactly can be queried but the interesting bits return *insufficient permission*.

If somebody else wants to give it a shot the manual might be a good start https://sentineldatahub.github.io/DataHubSystem/documents/DHuS%20Administration%20Manual_2.0.0.pdf

**Comment by gbaier** [comment]
I contacted the Sentinel data hub system team yesterday, about how to find out the installed DHuS version. Easy enough, one just has to add *api/stub/version* to the url, e.g.

1. https://scihub.copernicus.eu/dhus/api/stub/version
2. https://finhub.nsdc.fmi.fi/api/stub/version

I've added the functionality to get the version in the PR #342.

**Comment by j08lue** [comment]
Thanks @gbaier for researching this and submitting a PR. I am a bit worried about `sentinelsat` becoming a bit too dependent on all kinds of DHuS implementation details. Like inferring expected API functionality from version numbers. I would much rather see that we solve issues with upstream discrepancies by defensive programming - test whether the endpoint behaves like our reference implementation (SciHub) and if not, disable the functionality you are trying to reach (like LTA).

I am very much open for arguments why we should make an exception from that principle here. And @valgur and @Fernerkundung whether you support this principle.

**Comment by Fernerkundung** [comment]
Very much in favor of this:
>solve issues with upstream discrepancies by defensive programming 

This might fill our code with some verbose `try...except` blocks but I would argue is the Pythonic way to solve it.

This would also be a way to solve some of the inconsistencies in how a single end-point behaves - i.e. sudden behaviour changes of the LTA due to a bug.

**Comment by paalge** [comment]
I'm still hitting this error, though the fix works. Should it be added in a PR?

**Comment by valgur** [comment]
This issue should now be fixed by #469. Please re-open if that's not the case.

**Comment by men1n2** [comment]
@j08lue @valgur Thanks for the quick feedback. I see that a PR was open to solve the issue, is there still a need to open a separate issue (for archiving reasons) ?
